src/Dscp/DB/SQLite/Queries.hs

```diff
@@ -0,0 +1,313 @@
+        submissionSig  = sigSub^.ssWitness.swSig
+
+        student        = submission^.sStudentId
+        submissionHash = submission^.to hash
+        submissionType = submission^.sType
+        assignment     = submission^.sAssignment
+
+        course         = assignment^.aCourseId
+        assignmentHash = assignment^.to hash
+
+    transaction $ do
+        _ <- existsStudent student        `assert`     StudentDoesNotExist    student
+        _ <- existsCourse  course         `assert`     CourseDoesNotExist     course
+        _ <- getAssignment assignmentHash `assertJust` AssignmentDoesNotExist assignmentHash
+
+        _ <- isAssignedToStudent student assignmentHash
+            `assert` StudentWasNotSubscribedOnAssignment student assignmentHash
+
+        execute generateSubmissionRequest
+            ( submissionHash
+            , student
+            , assignmentHash
+            , submissionType
+            , submissionSig
+            )
+
+    return submissionHash
+  where
+    generateSubmissionRequest :: Query
+    generateSubmissionRequest = [q|
+        insert into  Submissions
+        values       (?, ?, ?, ?, ?)
+    |]
+
+setStudentAssignment :: DBM m => StudentId -> Hash Assignment -> m (StudentId, Hash Assignment)
+setStudentAssignment student assignment = do
+    execute setStudentAssignmentRequest (student, assignment)
+    return (student, assignment)
+  where
+    setStudentAssignmentRequest :: Query
+    setStudentAssignmentRequest = [q|
+        insert into  StudentAssignments
+        values      (?, ?)
+    |]
+
+isAssignedToStudent :: DBM m => StudentId -> Hash Assignment -> m Bool
+isAssignedToStudent student assignment = do
+    exists getStudentAssignmentQuery (student, assignment)
+  where
+    getStudentAssignmentQuery :: Query
+    getStudentAssignmentQuery = [q|
+        select  1
+        from    StudentAssignments
+        where   student_addr = ?
+           and  assignment_hash = ?
+    |]
+
+getGradesForCourseAssignments :: DBM m => StudentId -> CourseId -> m [PrivateTx]
+getGradesForCourseAssignments student course = do
+    query getGradesForCourseAssignmentsQuery (course, student)
+  where
+    getGradesForCourseAssignmentsQuery :: Query
+    getGradesForCourseAssignmentsQuery = [q|
+        select     Submissions.student_addr,
+                   Submissions.contents_hash,
+                   Assignments.course_id,
+                   Assignments.contents_hash,
+                   Assignments.desc,
+                   Submissions.signature
+                   grade
+                   time
+
+        from       Transactions
+
+        left join  Submissions
+               on  submission_hash = Submissions.hash
+
+        left join  Assignments
+               on  sssignment_hash = Assignments.hash
+              and  Assignments.course_id = ?
+
+        where      student_addr = ?
+    |]
+
+getStudentTransactions :: DBM m => PublicKey -> StudentId -> m [PrivateTx]
```

Comment:
I didn't quite get, why we need `PublicKey` here? The same question is for `getSignedSubmission`